working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,524 @@
+
+It is an error for a class to extend, implement, or mixin a type of the form
+`T?` for any `T`.
+
+It is an error for a class to extend, implement, or mixin the type `Never`.
+
+It is an error to call a method, setter, or getter on a receiver of static type
+`Never` (including via a null aware operator).
+
+It is an error to apply an expression of `Never` in the function position in a
+function call.
+
+It is an error if the static type of `e` in the expression `throw e` is
+potentially nullable.
+
+It is not an error for the body of a `late` field to reference `this`.
+
+It is an error for a formal parameter to be declared `late`.
+
+It is not a compile time error to write to a `final` variable if that variable
+is declared `late` and does not have an initializer.
+
+It is an error if the type `T` in the **on-catch** clause `on T catch` is
+potentially nullable.
+
+
+### Assignability
+
+The definition of assignability is changed as follows.  
+
+A type `T` is **assignable** to a type `S` if `T` is `dynamic`, or if `S` is a
+subtype of `T`.
+
+### Generics
+
+The default bound of generic type parameters is treated as `Object?`.
+
+### Type promotion, Definite Assignment, and Definite Completion
+
+**TODO** Fill this out.
+
+### Null promotion
+
+The machinery of type promotion is extended to promote the type of variables
+based on nullability checks subject to the same set of restrictions as normal
+promotion.  The relevant checks and the types they are considered to promote to
+are as follows.
+
+A check of the form `e == null` or of the form `e is Null` where `e` has static
+type `T` promotes the type of `e` to `Null` in the `true` continuation, and to
+**NonNull**(`T`) in the
+`false` continuation.
+
+A check of the form `e != null` or of the form `e is T` where `e` has static
+type `T?` promotes the type of `e` to `T` in the `true` continuation, and to
+`Null` in the `false` continuation.
+
+The static type of an expression `e!` is **NonNull**(`T`) where `T` is the
+static type of `e`.
+
+The **NonNull** function defines the null-promoted version of a type, and is
+defined as follows.
+
+- **NonNull**(Null) = Never
+- **NonNull**(_C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>) = _C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>  for class *C* other than Null (including Object).
+- **NonNull**(FutureOr<_T_>) = FutureOr<_T_>   
+- **NonNull**(_T_<sub>0</sub> Function(...)) = _T_<sub>0</sub> Function(...)
+- **NonNull**(Function) = Function
+- **NonNull**(Never) = Never
+- **NonNull**(dynamic) = dynamic
+- **NonNull**(void) = void   
+- **NonNull**(_X_) = X & **NonNull**(B), where B is the bound of X.
+- **NonNull**(_X_ & T) = X & **NonNull**(T)
+- **NonNull**(_T_?) = **NonNull**(_T_)
+- **NonNull**(_T_\*) = **NonNull**(_T_)
+
+#### Extended Type promotion, Definite Assignment, and Definite Completion
+
+These are extended as per separate proposal.
+
+### Runtime semantics
+
+#### Null assertion operator
```

Comment:
Would it be better to call this a "null check operator"?  Every time I hear "null assertion operator", I get momentarily confused because it sounds like it's an operator that asserts that its expression is null, whereas in fact it's the opposite.

Reply:
done.